cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -0,0 +1,1142 @@
+
+    newtype SigningKey ByronKey =
+           ByronSigningKey Byron.SigningKey
+      deriving newtype (ToCBOR, FromCBOR)
+
+    deterministicSigningKey :: AsType ByronKey -> Crypto.Seed -> SigningKey ByronKey
+    deterministicSigningKey AsByronKey seed =
+       ByronSigningKey (snd (Crypto.runMonadRandomWithSeed seed Byron.keyGen))
+
+    deterministicSigningKeySeedSize :: AsType ByronKey -> Word
+    deterministicSigningKeySeedSize AsByronKey = 32
+
+    getVerificationKey :: SigningKey ByronKey -> VerificationKey ByronKey
+    getVerificationKey (ByronSigningKey sk) =
+      ByronVerificationKey (Byron.toVerification sk)
+
+--    verificationKeyHash :: VerificationKey keyrole -> Hash keyrole
+
+--data instance Hash ByronKey = ByronKeyHash
+
+instance HasTextEnvelope (VerificationKey ByronKey) where
+    textEnvelopeType _ = "PaymentVerificationKeyByron"
+
+instance HasTextEnvelope (SigningKey ByronKey) where
+    textEnvelopeType _ = "SigningKeyByron"
+    -- TODO: fix these inconsistent names for the public testnet re-spin
+
+
+--
+-- Shelley payment keys
+--
+
+-- | Shelley-era payment keys. Used for Shelley payment addresses and witnessing
+-- transactions that spend from these addresses.
+--
+-- This is a type level tag, used with other interfaces like 'Key'.
+--
+data PaymentKey
+
+instance HasTypeProxy PaymentKey where
+    data AsType PaymentKey = AsPaymentKey
+    proxyToAsType _ = AsPaymentKey
+
+instance Key PaymentKey where
+
+    newtype VerificationKey PaymentKey =
+        PaymentVerificationKey (Shelley.VKey Shelley.Payment Shelley.TPraosStandardCrypto)
+      deriving newtype (ToCBOR, FromCBOR)
+
+    newtype SigningKey PaymentKey =
+        PaymentSigningKey (Shelley.SignKeyDSIGN Shelley.TPraosStandardCrypto)
+      deriving newtype (ToCBOR, FromCBOR)
+
+    deterministicSigningKey :: AsType PaymentKey -> Crypto.Seed -> SigningKey PaymentKey
+    deterministicSigningKey AsPaymentKey seed =
+        PaymentSigningKey (Crypto.genKeyDSIGN seed)
+
+    deterministicSigningKeySeedSize :: AsType PaymentKey -> Word
+    deterministicSigningKeySeedSize AsPaymentKey =
+        Crypto.seedSizeDSIGN proxy
+      where
+        proxy :: Proxy (Shelley.DSIGN Shelley.TPraosStandardCrypto)
+        proxy = Proxy
+
+    getVerificationKey :: SigningKey PaymentKey -> VerificationKey PaymentKey
+    getVerificationKey (PaymentSigningKey sk) =
+        PaymentVerificationKey (Shelley.VKey (Crypto.deriveVerKeyDSIGN sk))
+
+--    verificationKeyHash :: VerificationKey PaymentKey -> Hash PaymentKey
+--    verificationKeyHash (PaymentVerificationKey vk) =
+--        PaymentKeyHash (Shelley.hashKey vk)
+
+--newtype instance Hash PaymentKey =
+--    PaymentKeyHash (Shelley.KeyHash Shelley.Payment Shelley.TPraosStandardCrypto)
+
+instance HasTextEnvelope (VerificationKey PaymentKey) where
+    textEnvelopeType _ = "PaymentVerificationKeyShelley"
+    -- TODO: include the actual crypto algorithm name, to catch changes:
```

Comment:
Yeah, makes sense to do this.

Reply:
We could do it now except it'd break compatibility with the F&F testnet.